avoid scheduling jobs on compute nodes that are not cleaned up

garlick opened the pull request:
Problem: after a broker restart, "stuck" housekeeping, epilog, prolog, or job shell tasks might still be running, but flux is unaware and new work may be scheduled there even though there might be a problem, or those tasks might be holding on to resources.

When those things are run under systemd, we have the machinery for finding them and tracking them readily at hand. 

This PR does the following
- modifies `sdbus` so it can be loaded twice, once for user and once for system systemd instances
- add a `sdmon` monitoring module that tracks running flux units on both busses
- have `sdmon` join a `sdmon.idle` broker group at startup, after it has verified that no units are running
- modify the resource module so it monitors `sdmon.idle` instead of  `broker.online` when configured to use systemd

The net effect is that nodes that require cleanup remain offline until the node is cleaned up.

`sdmon` also logs the systemd units it has found.  Here's an example where I kill -9 a broker while housekeeping is running, then start it back up again
```
[  +0.192069] sdmon.err[7]: flux-housekeeping@fAMr3PwFUNb.service needs cleanup - resources are offline
...
[ +26.900866] sdmon.err[7]: cleanup complete - resources are online
```
Before cleanup is complete, `flux resource status` reports
```
      STATE UP NNODES NODELIST
       avail  ✔      7 picl[0-6]
      avail*  ✗      1 picl7
```

Seems like this does the bare minimum to resolve #6590 

This does seem a bit thin in the tooling department.   The possibilities are pretty broad so for now, I wanted to get this posted and get feedback on the way the resource module is tied into `sdmon` using broker groups.

grondo commented:
Wow, nice! I don't have any qualms with using an sdmon.idle broker group for the current implementation.

It seems like eventually we'd want various subsystems to be able to recapture their state from what sdmon has found. (for example the job execution system could reconnect to running jobs after restart, or terminate jobs that are not supposed to be running, the job manager could do something similar for prolog/epilog and housekeeping. Any thoughts on how that might work?  I realize bringing that up is a bit premature, but it could inform the solution here as a stepping stone. (I guess one thought is that as state is able to be recaptured, this would reduce the list of things that prevent a broker from joining `sdmod.idle`)

Also, since the `sdmon.idle` group is never left once joined (except in the case of a broker restart I'm assuming?), I'm wondering if there's a better term. I can't think of anything though, and it doesn't seem important enough to worry about it now (especially since it isn't really user/admin visible)

garlick commented:
Maybe `sdmon.online` is a better name for the group.  (It started out as an online group that was kept up to date and causes resource to post idle/busy events.  But then I realized it could be a lot simpler and didn't revisit the group name)

garlick commented:
Renamed the group and fixed a spelling error in a test caught in CI.

This still needs a test for the resource module portion of the proposed change so I'll leave it WIP for the moment.

garlick commented:
I added the missing test, so I'll drop the WIP.

One thing I should do before we merge this though is make  sure the systemd shipped with RHEL 8 allows sdbus to authetnicate to it with flux credentials.  I'll try to test that on fluke.

garlick commented:
Yep that worked
```
2025-02-07T16:41:43.412549-08:00 sdbus.info[0]: sd_bus_open_system: connected

```

garlick commented:
> (I guess one thought is that as state is able to be recaptured, this would reduce the list of things that prevent a broker from joining `sdmod.idle`)

Right, I like that way of thinking about it.

Hmm, we should also be trying to capture the state of any units that have completed but weren't reaped, and put that in a lost+found or something.  I need to refresh my memory on what happens to that state for the cases we're discussing here (the templated system units and transient user units).   That could be a follow-on PR.

But anyway, yeah, if a running unit can be reclaimed, we could let the node join `sdmon.online` before it terminates.

garlick commented:
Since #6662 is likely not feasible to get into our next release, it seems like this could be considered as is.

garlick commented:
Oh we should make sure this doesn't cause problems for
- #6652 

I will be offline for today but will double check that soon

grondo reviewed (approved):
This one looks good to me. I'm excited by the prospects of the `sdmon` module!

My only comment is that it would be nice if ranks offline due to sdmon could somehow bubble up the errors to a central location (e.g. something like `flux overlay errors`) so admins can easily tell the difference between "offline due to flux not started" vs "offline due to orphaned work". However, I'm not thinking of an easy way to provide that (or maybe there is already a way I'm not seeing) and it certainly shouldn't hold up this PR.

grondo commented:
> Oh we should make sure this doesn't cause problems for

What problems do you foresee there? The `shrink` should only affect jobs which aren't running `sdmon`, but I'm probably missing something obvious (Just thought maybe I can do some testing today while you're out)

garlick commented:
Just wanted to double check that this wasn't going to cause a problem:
> -  modify the resource module so it monitors sdmon.idle instead of broker.online when configured to use systemd

grondo commented:
I did install this PR branch onto my test cluster and performed some quick checks of `shrink` functionality and did not notice any effects. Since batch and alloc jobs won't have `systemd.enable` set, these instance won't even have the `sdmon` module loaded, so as far as `shrink` in the resource acquisition protocol, there's no effect.

I also tested that the resource module notices a rank suddenly disconnecting in the system instance, and this results in the expected behavior in the child. I wasn't really sure what else to check.

garlick commented:
I think this is OK since `sdmon.online` (formerly `sdmon.idle`) is only joined, never left.

garlick commented:
Good thought on adding something to `flux resource status`.  Maybe we could somehow annotate brokers that are online but not present in the `sdmon.online` set.  The lack of tools does seem to be the weak link here.

Another thought is to build a tool around `sdmon` that can do things like list orphaned units and signal them.

garlick commented:
Well, I'll go ahead and set MWP here to keep things moving along.  Thanks!
